### PR TITLE
istiod async changes

### PR DESCRIPTION
**Pull request by aryan16**
**Istiod Pilot Changes**
1. Update the Istiod pilot logic to make async calls to jwksuri and removed the calls from main flow.
2. Updated unit tests to support the new way of fetching public key from jwksuri

### CODE REVIEW

**Comment by howardjohn** [comment]
How does this related to the other PR? Seems the same?

File: pilot/pkg/model/jwks_resolver.go [hunk 1]
```diff
@@ -449,6 +448,7 @@ func (r *JwksResolver) refresh() bool {
 					atomic.AddUint64(&r.refreshJobFetchFailedCount, 1)
 					return
 				}
+				k.jwksURI = jwksURI
```
Comment (yangminzhu) [inline pilot/pkg/model/jwks_resolver.go:451]: you're not storing the key back to the cache.

**Comment by aryan16** [comment]
> How does this related to the other PR? Seems the same?

There were some issues with other branch , so I created this new one. Will close my older PR.

File: pilot/pkg/model/jwks_resolver.go [hunk 2]
```diff
@@ -384,41 +373,49 @@ func (r *JwksResolver) refresher() {
...
-			}
-			lastHasError = currentHasError
-			r.refreshTicker.Reset(r.refreshInterval)
+			lastHasError = r.refreshCache(lastHasError)
 		case <-closeChan:
 			r.refreshTicker.Stop()
 			return
+		case jwtKeyData := <-jwksuriChannel:
+			lastHasError = r.refreshCache(lastHasError)
+			if lastHasError {
```
Comment (yangminzhu) [inline pilot/pkg/model/jwks_resolver.go:382]: I think you do not need to log this anymore, it should already be logged inside the refresh().

**Comment by howardjohn** [comment]
I see. It would be ideal to just force push to that one to keep history of the review. You can do `git checkout broken-branch; git reset --hard istiod-sync` and it will make the old branch the same as this one. Then `git push origin --force-with-lease istiod-sync`

File: pilot/pkg/model/jwks_resolver.go (hunk 1 above)
Comment (aryan16) [reply pilot/pkg/model/jwks_resolver.go:451]: we are storing it here - https://github.com/istio/istio/blob/master/pilot/pkg/model/jwks_resolver.go#L462. Do you think there's a point to store it if we can't fetch the pubkey from that ?

**Comment by aryan16** [comment]
> I see. It would be ideal to just force push to that one to keep history of the review. You can do `git checkout broken-branch; git reset --hard istiod-sync` and it will make the old branch the same as this one. Then `git push origin --force-with-lease istiod-sync`

Tried to follow this, but that branch has added all the previous commits which made it look really hard to analyze. I will add that PR here for reference.

File: pilot/pkg/security/authn/v1beta1/policy_applier_test.go [hunk 3]
```diff
@@ -691,18 +691,18 @@ func TestJwtFilter(t *testing.T) {
...
-		model.JwtPubKeyRefreshIntervalOnFailure, model.JwtPubKeyRetryInterval)
-	defer push.JwtKeyResolver.Close()
-
-	push.ServiceIndex.HostnameAndNamespace[host.Name("jwt-token-issuer.mesh")] = map[string]*model.Service{}
-	push.ServiceIndex.HostnameAndNamespace[host.Name("jwt-token-issuer.mesh")]["mesh"] = &model.Service{
-		Hostname: "jwt-token-issuer.mesh.svc.cluster.local",
-	}
 	for _, c := range cases {
 		t.Run(c.name, func(t *testing.T) {
+			push := model.NewPushContext()
```
[the remark this answers is not in the document]
Comment (aryan16) [reply pilot/pkg/security/authn/v1beta1/policy_applier_test.go:696]: Since we are fetching `pubkey` in the background, the results could be flaky if we create one single object and use it for every test case becuase we don't know when the key will be ready. It's a safe bet to use it inside the for loop and use the `fakePubKey` for every scenario.

File: pilot/pkg/model/jwks_resolver.go (hunk 1 above)
Comment (yangminzhu) [reply pilot/pkg/model/jwks_resolver.go:437]: the problem is the old key with empty jwksURI is not removed, you're not updating it but just adding a new key, by update I mean you should replace the old one with the new content.

File: pilot/pkg/security/authn/v1beta1/policy_applier_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (yangminzhu) [reply pilot/pkg/security/authn/v1beta1/policy_applier_test.go:696]: could you clarify why it is flaky to use one object for every test cases? Isn't that how it is working in real cases? The test should not deviate from the real behavior.

File: pilot/pkg/model/jwks_resolver.go [hunk 4]
```diff
@@ -232,51 +233,32 @@ func (r *JwksResolver) GetPublicKey(issuer string, jwksURI string) (string, erro
 		e.lastUsedTime = now
 		r.keyEntries.Store(key, e)
 		if e.pubKey == "" {
-			return e.pubKey, errEmptyPubKeyFoundInCache
+			return e.pubKey
```
Comment (yangminzhu) [inline pilot/pkg/model/jwks_resolver.go:236]: there is already a return below, you can just remove the if-statement together.

File: pilot/pkg/security/authn/v1beta1/policy_applier_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (aryan16) [reply pilot/pkg/security/authn/v1beta1/policy_applier_test.go:696]: all the tests cases separately call the `buildJwksUri()` which calls the `GetOrUpdatePubKey()`. So for the first case it will always return fakejwks, but the other cases with same `jwksuri` and `issuer` will have uncertainty as the method will return fakejwks when cache is empty but if cache got filled in any of the previous cases it will return the real pub key. So we don't exactly know when it will return real pubkey.

**Comment by aryan16** [comment]
cc : @yangminzhu, @howardjohn

File: pilot/pkg/security/authn/v1beta1/policy_applier_test.go [hunk 5]
```diff
@@ -705,6 +704,15 @@ func TestJwtFilter(t *testing.T) {
 	for _, c := range cases {
 		t.Run(c.name, func(t *testing.T) {
 			istiotest.SetBoolForTest(t, &features.EnableRemoteJwks, c.enableRemoteJwks)
+			NewPolicyApplier("root-namespace", c.in, nil, push).JwtFilter()
+			for _, config := range c.in {
+				for _, jwtRule := range config.Spec.(*v1beta1.RequestAuthentication).JwtRules {
+					retryToCheckCache(5, 20*time.Millisecond, func() bool {
+						return push.JwtKeyResolver.CheckPubKeyExistInCache(jwtRule.Issuer, jwtRule.JwksUri)
+					})
```
Comment (howardjohn) [inline pilot/pkg/security/authn/v1beta1/policy_applier_test.go:712]:
```suggestion
			retry.UntilOrFail(t, func() bool {
						return push.JwtKeyResolver.CheckPubKeyExistInCache(jwtRule.Issuer, jwtRule.JwksUri)
					})
```

and can update other methods + remove the retryToCheckCache function

File: pilot/pkg/security/authn/v1beta1/policy_applier_test.go [hunk 6]
```diff
@@ -705,6 +704,15 @@ func TestJwtFilter(t *testing.T) {
...
+				}
+			}
+
 			if got := NewPolicyApplier("root-namespace", c.in, nil, push).JwtFilter(); !reflect.DeepEqual(c.expected, got) {
```
Comment (howardjohn) [inline pilot/pkg/security/authn/v1beta1/policy_applier_test.go:716]: should we use the `JwtFilter` from L707? Or if we are testing its not changed, then check both times?

File: pilot/pkg/security/authn/v1beta1/policy_applier_test.go (hunk 5 above)
Comment (howardjohn) [reply pilot/pkg/security/authn/v1beta1/policy_applier_test.go:712]: Actually its not clear what this function was doing originally... since we do not do anything if/when it times out?

File: pilot/pkg/security/authn/v1beta1/policy_applier_test.go [hunk 7]
```diff
@@ -405,7 +405,7 @@ func TestJwtFilter(t *testing.T) {
 									JwksSourceSpecifier: &envoy_jwt.JwtProvider_LocalJwks{
 										LocalJwks: &core.DataSource{
 											Specifier: &core.DataSource_InlineString{
-												InlineString: test.JwtPubKey1,
+												InlineString: test.JwtPubKey2,
```
Comment (howardjohn) [inline pilot/pkg/security/authn/v1beta1/policy_applier_test.go:408]: why is this changed?

File: pilot/pkg/model/jwks_resolver.go [hunk 8]
```diff
@@ -220,63 +222,40 @@ func newJwksResolverWithCABundlePaths(
...
-		if err != nil {
-			log.Errorf("Failed to fetch public key from %q: %v", jwksURI, err)
-		}
-		pubKey = string(resp)
-	}
-
 	r.keyEntries.Store(key, jwtPubKeyEntry{
-		pubKey:            pubKey,
+		pubKey:            "",
 		lastRefreshedTime: now,
```
Comment (howardjohn) [inline pilot/pkg/model/jwks_resolver.go:239]: it hasn't actually be refreshed 'now', so this seems inaccurate?

File: pilot/pkg/model/jwks_resolver.go [hunk 9]
```diff
@@ -449,13 +431,17 @@ func (r *JwksResolver) refresh() bool {
...
+				r.keyEntries.Delete(k)
+				k.jwksURI = jwksURI
 			}
-
 			resp, err := r.getRemoteContentWithRetry(jwksURI, networkFetchRetryCountOnRefreshFlow)
 			if err != nil {
 				hasErrors = true
 				log.Errorf("Failed to refresh JWT public key from %q: %v", jwksURI, err)
 				atomic.AddUint64(&r.refreshJobFetchFailedCount, 1)
+				if oldPubKey == "" {
```
Comment (howardjohn) [inline pilot/pkg/model/jwks_resolver.go:442]: when this happens `compareJWKSResponse` will spam a confusing error now

File: pilot/pkg/model/jwks_resolver.go [hunk 10]
```diff
@@ -220,63 +222,40 @@ func newJwksResolverWithCABundlePaths(
...
-			// TODO(xulingqing): Find a better way to reject the request without using the fake jwks.
-			jwtPubKey = CreateFakeJwks(jwksURI)
-		}
+		jwtPubKey = r.GetOrUpdatePublicKey(jwtIssuer, jwksURI)
+	}
+	if jwtPubKey == "" {
+		log.Infof("The JWKS key is not yet fetched for issuer %s (%s), using a fake JWKS for now", jwtIssuer, jwksURI)
+		// This is a temporary workaround to reject a request with JWT token by using a fake jwks when istiod failed to fetch it.
+		// TODO(xulingqing): Find a better way to reject the request without using the fake jwks.
+		jwtPubKey = CreateFakeJwks(jwksURI)
```
Comment (howardjohn) [inline pilot/pkg/model/jwks_resolver.go:258]: Are we sure that this is the right user facing behavior?

This means that if I apply some PeerAuthentication, my traffic will immediately break for a bit before it fetches later?

File: pilot/pkg/model/jwks_resolver.go [hunk 11]
```diff
@@ -93,6 +92,9 @@ var (
 
 	// JwtPubKeyRefreshInterval is the running interval of JWT pubKey refresh job.
 	JwtPubKeyRefreshInterval = features.PilotJwtPubKeyRefreshInterval
+
+	// channel for making jwksuri request aynsc
+	jwksuriChannel = make(chan jwtKey, 5)
```
Comment (howardjohn) [inline pilot/pkg/model/jwks_resolver.go:97]: Slightly worried about size of 5.. users probably don't have many more but if they do it can block everything

File: pilot/pkg/model/jwks_resolver.go [hunk 12]
```diff
@@ -365,6 +372,7 @@ func (r *JwksResolver) getRemoteContentWithRetry(uri string, retry int) ([]byte,
 
 	for i := 0; i < retry; i++ {
 		body, err := getPublicKey()
+		fmt.Println(r.retryInterval)
```
Comment (yangminzhu) [inline pilot/pkg/model/jwks_resolver.go:375]: This should be deleted from the PR.

Comment (aryan16) [reply pilot/pkg/model/jwks_resolver.go:375]: oh yeah, I forgot to remove this. Will do it.

File: pilot/pkg/security/authn/v1beta1/policy_applier_test.go [hunk 13]
```diff
@@ -486,14 +486,15 @@ func TestJwtFilter(t *testing.T) {
 			},
 		},
 		{
-			name: "JWT policy with bad Jwks URI",
```
Comment (aryan16) [inline pilot/pkg/security/authn/v1beta1/policy_applier_test.go:489]: Changed the position of `bad jwks uri` test because it initiates the background fetch and since we are trying to fetch `pubkey` for all the jwksuri in the cache we would update the key for other jwksuri (as our mock server returns pubkey1 in first hit and pubkey2 for rest of the hits). If we don't do this, it would create some flakiness in the code.

File: pilot/pkg/model/jwks_resolver.go [hunk 14]
```diff
@@ -259,20 +264,22 @@ func (r *JwksResolver) GetPublicKey(issuer string, jwksURI string) (string, erro
 		lastRefreshedTime: now,
 		lastUsedTime:      now,
 	})
-
+	if err != nil {
+		// fetching the public key in the background
+		jwksuriChannel <- key
```
[the remark this answers is not in the document]
Comment (aryan16) [reply pilot/pkg/model/jwks_resolver.go:269]: It can be done in concurrent execution, but in that case we need to change `r.refreshInterval` (which is 20 mins rn) to some smaller value to make sure that our ticker executes instantly. If we do that there might be other cases which could go wrong.

[the remark this answers is not in the document]
Comment (aryan16) [reply pilot/pkg/model/jwks_resolver.go:269]: what type of edge cases you're talking about ?

[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/model/jwks_resolver.go:269]: oops, misunderstood ticker. Replace what I said with `r.refreshTicker.C <- time.Now()`.

The issue would be if the refreshTicker select block is running and slow (it could be making 100s of http calls). We do the send on jwksuriChannel >5 times, it will block until the refreshTicker part is done.

Alternatively we can just have the on-demand async calls decoupled entirely from the refreshTicker (and probably just refresh that 1 url, not loop over all of them

[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/model/jwks_resolver.go:269]: Actually the single call may be better, otherwise for every new jwks we need to refreshall other jwks - may not scale well

[the remark this answers is not in the document]
Comment (yangminzhu) [reply pilot/pkg/model/jwks_resolver.go:269]: > otherwise for every new jwks we need to refreshall other jwks - may not scale well

yes, we should address this.

The other way could be to update the refresh() logic so that it only fetches the jwks URI that are newly added from here.

**Review by yangminzhu — COMMENTED** [review]
Overall looks good, please also address the remaining open comments from john.

File: pilot/pkg/model/jwks_resolver.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (aryan16) [reply pilot/pkg/model/jwks_resolver.go:269]: I have discussed with him, since we don't have many `jwksuri` in the cache in one single time. I will add a new logic in my follow-up work where we will just try to fetch the pubkey only for the jwksuri which is failing.

**Comment by yangminzhu** [comment]
@aryan16 there are still many open comments in the PR, I'm not sure if they are already resolved or not needed, could you update in each comment to indicate if it's resolved or not needed?

File: pilot/pkg/model/jwks_resolver.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (yangminzhu) [reply pilot/pkg/model/jwks_resolver.go:269]: Could you just do it in this PR? I'm a bit worried about the performance even if the customer is only using a small number of `jwksUri` because each of the `jwksUri` will trigger a new config push right?

[the remark this answers is not in the document]
Comment (aryan16) [reply pilot/pkg/model/jwks_resolver.go:269]: even after we refresh one single `jwksuri`, it is gonna create a new config push and will push the whole cache again. The only drawback which I see in our current solution is that it will try to refresh all the `jwksuri` in the cache. 
@howardjohn do we have any option to push one single `jwksuri` - `pubkey` to envoy ?

**Comment by aryan16** [comment]
> @aryan16 there are still many open comments in the PR, I'm not sure if they are already resolved or not needed, could you update in each comment to indicate if it's resolved or not needed?

sure, I forgot to close few comments

**Comment by aryan16** [comment]
@yangminzhu @howardjohn PTAL

File: pilot/pkg/model/jwks_resolver.go [hunk 15]
```diff
@@ -384,41 +394,53 @@ func (r *JwksResolver) refresher() {
...
+			if !r.jwksUribackgroundChannel {
+				lastHasError = r.refreshCache(lastHasError)
 			}
-			lastHasError = currentHasError
-			r.refreshTicker.Reset(r.refreshInterval)
 		case <-closeChan:
 			r.refreshTicker.Stop()
 			return
+		case <-jwksuriChannel:
+			r.jwksUribackgroundChannel = true
```
Comment (howardjohn) [inline pilot/pkg/model/jwks_resolver.go:404]: style: maybe we can just make `jwksUribackgroundChannel` a parameter to `refreshCache`? Using the struct level variable makes it pretty hard to follow

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/model/jwks_resolver.go (hunk 15 above)
Comment (aryan16) [reply pilot/pkg/model/jwks_resolver.go:404]: With the struct level variable we can stop the main ticker execution when our background execution is going on. This would resolve the consistency issues, added a check on line no. 397
